framework/merchant_api/infrastructure/merchant_data_providers/coles_provider.py

```diff
@@ -0,0 +1,127 @@
+
+from domain.entities.merchant import Merchant
+from framework.merchant_api.domain.entities.coles_product_offer import \
+    ColesProductOffer
+from framework.merchant_api.domain.entities.dora_product import DoraProduct
+from framework.merchant_api.domain.entities.scraped_product_offer import \
+    ScrapedProductOffer
+from framework.merchant_api.domain.enumerations.supported_merchant import \
+    SupportedMerchant
+from framework.merchant_api.infrastructure.session import get_cached_session
+from framework.merchant_api.services.imerchant_data_provider import \
+    IMerchantDataProvider
+
+
+class ColesProvider(IMerchantDataProvider):
+
+    #region ---------------- Fields ----------------
+
+    _is_healthy = True
+
+    #endregion Fields
+
+    #region ---------------- Properties ----------------
+
+    @property
+    def base_url(self) -> str:
+        return 'https://www.coles.com.au'
+
+    @property
+    def is_healthy(self) -> bool:
+        return self._is_healthy
+
+    @is_healthy.setter
+    def is_healthy(self, val: bool) -> None:
+        self._is_healthy = val
+
+    @property
+    def priority(self) -> int:
+        return 1
+
+    @property
+    def supported_merchants(self) -> List[str]:
+        return [
+            SupportedMerchant.COLES
+        ]
+
+    #endregion Properties
+
+    #region ---------------- Methods ----------------
+
+    def get_product(self, product: DoraProduct) -> ScrapedProductOffer:
+        with get_cached_session() as _Session:
+            _Response = _Session.get('https://www.coles.com.au/')
+            _Soup = BeautifulSoup(_Response.text, features="html.parser")
+            _BuildID = json.loads(_Soup.find(id='__NEXT_DATA__').contents[0])['buildId']
+
+            _Url = f'https://www.coles.com.au/_next/data/{_BuildID}/en/search.json'
+            _Params = {
+                'q': product.name,
+                'page': 1
+            }
+
+            _PageSearchResult = _Session.get(_Url, _Params).json()['pageProps']['searchResults']
+            for _ProductSearchResult in _PageSearchResult['results']:
+                if _ProductSearchResult['_type'] == "PRODUCT" and str(_ProductSearchResult["id"]) == product.merchant_stockcode:
+                    return self._translate_offer(ColesProductOffer.model_validate(_ProductSearchResult))
+
+    def search_by_term(self, search_term: str, merchant: Merchant, result_limit: int) -> List[ScrapedProductOffer]:
+        with get_cached_session() as _Session:
+            _Response = _Session.get('https://www.coles.com.au/')
+            _Soup = BeautifulSoup(_Response.text, features="html.parser")
+            _BuildID = json.loads(_Soup.find(id='__NEXT_DATA__').contents[0])['buildId']
+
+            _Url = f'https://www.coles.com.au/_next/data/{_BuildID}/en/search.json'
+            _Params = {
+                'q': search_term,
+                'page': 1
+            }
+
+            _ScrapedProductOffers = []
+            while True:
```

Comment:
for all providers, may want to add safety net for infinite spam, maybe a time based back off?